staging/src/k8s.io/sample-controller/controller.go

```diff
@@ -0,0 +1,301 @@
+			}
+		},
+	})
+	c.deploymentsInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: c.handleObject,
+		UpdateFunc: func(old, new interface{}) {
+			if !reflect.DeepEqual(old, new) {
+				c.handleObject(new)
+			}
+		},
+		DeleteFunc: c.handleObject,
+	})
+
+	glog.Info("Starting informer factories")
+	go c.informerFactory.Start(stopCh)
+	go c.exampleInformerFactory.Start(stopCh)
+
+	glog.Info("Waiting for informer caches to sync")
+	if ok := cache.WaitForCacheSync(stopCh, c.nginxInformer.Informer().HasSynced, c.deploymentsInformer.Informer().HasSynced); !ok {
+		return fmt.Errorf("failed to wait for caches to sync")
+	}
+
+	glog.Info("Starting workers...")
+	// We use a WaitGroup here so that upon initial shutdown signal, we wait
+	// for the items that are currently being processed to finish processing
+	var wg sync.WaitGroup
+	for i := 0; i < 2; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			c.worker()
+		}()
+	}
+
+	glog.Info("Started workers")
+	<-stopCh
+	glog.Info("Shutting down workers")
+	c.workqueue.ShutDown()
+	wg.Wait()
+	glog.Info("Workers shutdown")
+
+	return nil
+}
+
+func (c *Controller) worker() {
+	for {
+		obj, shutdown := c.workqueue.Get()
+
+		if shutdown {
+			return
+		}
+
+		err := func(obj interface{}) error {
+			defer c.workqueue.Done(obj)
+			var key string
+			var ok bool
+			if key, ok = obj.(string); !ok {
+				c.workqueue.Forget(obj)
+				runtime.HandleError(fmt.Errorf("expected string in workqueue but got %#v", obj))
+				return nil
+			}
+			if err := c.syncHandler(key); err != nil {
+				return fmt.Errorf("error syncing '%s': %s", key, err.Error())
+			}
+			c.workqueue.Forget(obj)
+			return nil
+		}(obj)
+
+		if err != nil {
+			runtime.HandleError(err)
+			continue
+		}
+	}
+}
+
+func (c *Controller) syncHandler(key string) error {
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		runtime.HandleError(fmt.Errorf("invalid resource key: %s", key))
+		return nil
+	}
+
+	nginx, err := c.nginxInformer.Lister().NGINXs(namespace).Get(name)
+
+	if err != nil {
+		if errors.IsNotFound(err) {
+			runtime.HandleError(fmt.Errorf("nginx '%s' in work queue no longer exists", key))
+			return nil
+		}
+
+		return err
+	}
+
+	deploymentName := nginx.Spec.DeploymentName
+	if deploymentName == "" {
+		// we don't return an error here so that the item doesn't get requeued
+		// until it is next modified
+		runtime.HandleError(fmt.Errorf("%s: deployment name must be specified", key))
+		return nil
+	}
+
+	deployment, err := c.deploymentsInformer.Lister().Deployments(nginx.Namespace).Get(deploymentName)
+	if errors.IsNotFound(err) {
+		deployment, err = c.clientset.AppsV1beta1().Deployments(nginx.Namespace).Create(newDeployment(nginx))
+	}
+
+	if err != nil {
+		return err
+	}
+
+	return c.updateNGINXStatus(nginx, deployment)
+}
+
+func (c *Controller) updateNGINXStatus(nginx *example.NGINX, deployment *apps.Deployment) error {
+	update := nginx.DeepCopy()
```

Comment:
nit: maybe `nginxCopy` instead of `update`?

Reply:
👍  done